### DISCUSSION TITLE
Allowing per Pod / Datacenter directories

### DISCUSSION THREAD

**Discussion by Serazio** [aristanetworks/avd#3233] [Ideas]
Hi Guys,

before writing a feature request, i would like to discuss an idea. 

For our new multi pod multi datacenter design i'm exploring avd for automation currently.  Especially in large environments which hundreds of switches and multiple datacenters / pods, having all generated intended, structured configs and documentation in  single per type directories harms in clarity. 

For example in https://github.com/aristanetworks/ansible-avd/blob/cfd50d669f320e300635a10b3fb026ec741f8ef5/ansible_collections/arista/avd/roles/eos_designs/tasks/main.yml#L17C1-L28C1 the task for creating the directory is set to run_once. 

As it would reduce folder sizes in large envs, I would like to suggest, to run it per device or better per pod/dc( don't know if this could work ), to allow something like this per pod:

```---
### group_vars/DC1_POD1.yml

dc_name: DC1
pod_name: POD1

output_dir_name: 'intended/{{dc_name}}/{{pod_name}}'
structured_dir_name: 'structured_configs/{{dc_name}}/{{pod_name}}'
eos_config_dir_name: 'configs/{{dc_name}}/{{pod_name}}'
devices_dir_name: 'devices/{{dc_name}}/{{pod_name}}'
post_running_config_backup_dir_name: 'config_backup/{{dc_name}}/{{pod_name}}'
pre_running_config_backup_dir_name: 'config_backup/{{dc_name}}/{{pod_name}}'
```

Which should result in a directory structure like this:
```
intented
- DC1
- - POD1
- - - leaf1.cfg
- - - leaf2.cfg
- DC2
- - POD2
- - - leaf1.cfg
- - - leaf2.cfg
...
structured_configs
- DC1
- - POD1
- - - leaf1.cfg
- - - leaf2.cfg
- DC2
- - POD2
- - - leaf1.cfg
- - - leaf2.cfg
...
```

Maybe it could be a solution to implement a new option `structured_directorys: true/false` which allows to select either the current way, or my above idea.  

Regards
Alex

**Comment by ClausHolbechArista** [comment]
Hi Alex

Thank you for sharing your idea.

Just for clarity, you are already producing the files across these folders today right? (At least I see no issue in doing so as long as you create the directories yourself)

So the ask/idea is to create the directories automatically so when you add a POD it won't break right?

I think the best way to implement this would be to do it inside the action plugins `eos_designs_structured_config` and `validate_and_template`, which are the two providing per-device outputs. We could simply check and create the directories automatically as needed. It would replace the existing task.

Let me know if I understood you correctly.

Thanks
Claus

**Reply by Serazio** [reply]
Hi Claus,

i do not create those folders currently, due to the "folder does not exist" error while building the configurations. Additionally my test setup is not so big at the moment. 

But yes, you are absolutely right :) 

Regards
Alex
